Enhance README and CMakeLists for OpenCV module customization

abdelaziz-mahdy opened the pull request:
should fix https://github.com/rainyl/opencv_dart/issues/313, didnt try it yet though, but the yaml option was annyoing to implement and found out this option will make it easier to use, if it looks good on your side we can add an example to test it i guess

This pull request introduces a new feature to allow customization of OpenCV modules in the `opencv_dart` and `opencv_core` packages. It adds support for a user-defined configuration file (`opencv_config.cmake`) to enable or disable specific OpenCV modules during the build process. The changes include updates to documentation and build scripts to support this feature.

Edit by rainyl:
TODO:
- [x] dart parser script
- [x] opencv_core ?
  - [x] CMakeLists.txt
  - [x] readme
  - [x] example
- [x] opencv_dart
  - [x] CMakeLists.txt
  - [x] readme
  - [x] example

abdelaziz-mahdy committed:
Enhance README and CMakeLists for OpenCV module customization

- Added documentation on customizing OpenCV modules via `opencv_config.cmake` in both `opencv_core` and `opencv_dart` READMEs.
- Updated `CMakeLists.txt` in both packages to include user-defined configurations if `opencv_config.cmake` exists, falling back to default settings otherwise.

abdelaziz-mahdy commented:
also i dont know how to handle that in case of macos and ios :)

rainyl commented:
> should fix #313, didnt try it yet though, but the yaml option was annyoing to implement and found out this option will make it easier to use, if it looks good on your side we can add an example to test it i guess
> 
> This pull request introduces a new feature to allow customization of OpenCV modules in the `opencv_dart` and `opencv_core` packages. It adds support for a user-defined configuration file (`opencv_config.cmake`) to enable or disable specific OpenCV modules during the build process. The changes include updates to documentation and build scripts to support this feature.

@abdelaziz-mahdy the solution is doable but also dangerous, `include()` means users can do anything including deletion and malicious code injection, although they do have full control of the package source code, but this will make it easier to attack.

Actually even though #313 is opened, no many demands received for now, and just like you mentioned, it will be solved by native-assets, therefore I didn't active to solve it.

> also i dont know how to handle that in case of macos and ios :)

This will be easier, we can only provide `'DartCvIOS/core'` in `opencv_core.podspec` and require users to explicitly provide needed modules inside the podfile.

```ruby
# <project>/ios/Podfile

target 'Runner' do
  use_frameworks!
  use_modular_headers!

  flutter_install_all_ios_pods File.dirname(File.realpath(__FILE__))
  target 'RunnerTests' do
    inherit! :search_paths
  end
  pod 'DartCvIOS/imgproc' '4.11.0.2' # add this
end

# ...
```

So in summary, if we want to support the optional modules from now, I would also prefer parsing the specific options in `pubspec.yaml` following the format like (same as dart user-defines https://github.com/dart-lang/native/issues/39) 
```yaml
hooks:
  user_defines:
    dartcv4:
      exclude_modules:
        - contrib
        - dnn
        - features2d
```

which means we need to parse yaml or at least filter the needed options in CMakelists.txt.

abdelaziz-mahdy commented:
Oh actually I really didn't consider the malicious intents in my implementation 


Yaml will require parsing and generation of cmake options based on it 


Will check if I am able to do it or not, but I didn't understand the iOS/macos parts

rainyl commented:
> Yaml will require parsing and generation of cmake options based on it

Yes, you can try to ask GPT or claude from some help.

> but I didn't understand the iOS/macos parts

currently, the full dependency of `DartCvIOS` is specified inside [opencv_core.podspec](https://github.com/rainyl/opencv_dart/blob/033b500a6c79e2b7123ff075b5119243077f750e/packages/opencv_core/ios/opencv_core.podspec#L30-L31), but I have split the modules into several parts in  [DartCvIOS](https://github.com/rainyl/dartcv/blob/7c0a54a6273fe3b0af8f3a501d116cb912bcc0a6/DartCvIOS.podspec#L77-L149), which means, users can just add some modules to their `Podfile` to introduce them, this can be done inside the `<flutter_project>/ios/Podfile` like I mentioned before to avoid unnecessary modules, of course, in this way, we need to change the `opencv_core.podspec` to replace the full dependency of `DartCvIOS` with `DartCvIOS/core` (core module is always needed), similar to macos.

abdelaziz-mahdy committed:
Refactor OpenCV module customization to use pubspec.yaml

- Updated README files for both `opencv_core` and `opencv_dart` to reflect the new method of customizing OpenCV modules via the `pubspec.yaml` file instead of `opencv_config.cmake`.
- Introduced a Dart script to generate `dartcv_modules.cmake` based on the specified modules in `pubspec.yaml`.
- Modified `CMakeLists.txt` in both packages to include the generated `dartcv_modules.cmake`, ensuring proper module configuration during the build process.

abdelaziz-mahdy commented:
I played around with ai/cursor to create the yaml parser and integrate it, so let me know if it makes sense now or not, and i cant see the logs when building with cmake, so i really dont know if its working or not (also i think android examples needs to be updated for gradle)

rainyl commented:
> I played around with ai/cursor to create the yaml parser and integrate it, so let me know if it makes sense now or not, and i cant see the logs when building with cmake, so i really dont know if its working or not

looks good, will test it. you can pass -v option when running `flutter build` and the detailed logs will be shown, also you can use `flutter build apk -v > build.log` to redirect them to a file if it's too long. :)

> also i think android examples needs to be updated for gradle

Feel free to do it.

abdelaziz-mahdy committed:
added default fallback incase target-platform is not set, and upgraded gradle versions

abdelaziz-mahdy committed:
Update Android Gradle plugin and Gradle wrapper versions

- Upgraded Android Gradle plugin from version 8.1.0 to 8.10.0.
- Updated Gradle wrapper distribution from version 8.5 to 8.11.1.

rainyl committed:
add gen_cmake_vars.dart for dartcv4, make CMakeLists.txt run the script to get module configurations

rainyl committed:
dart format ✅

rainyl commented:
I just got some time and tested it, finding that your current implementation is okay but not elegant enough, e.g., cmake can't get user's project directory and flutter doesn't provide a concrete way to get it, and it will be better if the generation step doesn't produce any files.

Therefore, I have made some changes for `dartcv4` and `opencv_core`, basically add a yaml parser inside `packages/dartcv/bin`, making it output the constructed string and calling the command inside `CmakeLists.txt` like you did, and parsing the output string using Regexp and set needed variables, in this way, no output files needed and will be safer and cleaner. 

The current implementation works on android but need to be tested on linux and windows, you can try to do it if you have some time, if not I will do it, maybe in the next week if I have some free time, too busy recently :(

rainyl committed:
fix cmakelists.txt

abdelaziz-mahdy commented:
Thank you for letting me know the changes, yeah I thought about dealing with the paths and file generated will be annoying and as you mentioned before maybe someone will inject commands into it 

This is why I went with the string approach, and I don't have Linux nor windows machines so I can't test it sadly 🥹

Also thank you and take your time I don't want the feature currently but I liked the addition and thought it will be great for users experience

rainyl commented:
I have edited the PR to add a todo list, hope you won't mind it.

Also, if the optional modules supported, I am considering whether it's necessary to keep the package `opencv_core`, since the only difference between `opencv_core` and `opencv_dart` is the supported modules, maybe we can deprecate `opencv_core`?

rainyl commented:
> This is why I went with the string approach, and I don't have Linux nor windows machines so I can't test it sadly 🥹

Aha, it's okay, I will test it sometime.

rainyl committed:
fix windows

rainyl committed:
dart format ✅

rainyl committed:
fix linux

rainyl committed:
bump dartcv4 to 1.1.4, fix android

rainyl commented:
@abdelaziz-mahdy Now, the optional module feature is supported for Android, Linux and Windows, since cocoapods is used for ios and macos, I didn't find a solution to access the pubspec.yaml file of user's project inside `opencv_dart.podspec`, so it seems the only solution is letting users explicitly set required dependencies inside their `Podfile`.

However, this also means we need to change the default dependency inside `opencv_dart.podspec` from 
```ruby
s.dependency 'DartCvMacOS', "#{dartcv_version}"
s.dependency 'DartCvMacOS/dnn', "#{dartcv_version}"
s.dependency 'DartCvMacOS/videoio', "#{dartcv_version}"
```

to 
```ruby
s.dependency 'DartCvMacOS/core', "#{dartcv_version}"
```

and then users can set 
```ruby
target 'Runner' do
  use_frameworks!
  use_modular_headers!

  flutter_install_all_macos_pods File.dirname(File.realpath(__FILE__))
  target 'RunnerTests' do
    inherit! :search_paths
  end

  pod 'DartCvIOS/imgproc' '4.11.0.2' # add this
end
```

To be honest, it's a breaking change and may affect many existed projects, also it's not elegant enough, so I prefer to remain the current status for ios and macos, what do you think?

rainyl committed:
update readme and changelog

abdelaziz-mahdy commented:
I think due to the nature of it, if there is a way we can make it the same and users somehow can exclude some libs that will be great, 
Other than that.

A breaking change is not worth it in my mind (iOS/macos) can remain big until native assets is released.

rainyl commented:
Thanks 😄 

Then I will merge the PR, thanks for your kind suggestions and contributions~
